contracts/finance/Vesting.sol

```diff
@@ -0,0 +1,284 @@
+
+        for (uint256 i = 0; i < ids_.length; i++) {
+            beneficiaryVestings_[i] = _vestings[ids_[i]];
+        }
+
+        return beneficiaryVestings_;
+    }
+
+    function getVestingIds(address beneficiary_) public view virtual returns (uint256[] memory) {
+        return _beneficiaryIds[beneficiary_].values();
+    }
+
+    function getWithdrawableAmount(uint256 vestingId_) public view virtual returns (uint256) {
+        VestingData storage _vesting = _vestings[vestingId_];
+        Schedule storage _schedule = _schedules[_vesting.scheduleId];
+
+        return _getWithdrawableAmount(_vesting, _schedule, block.timestamp);
+    }
+
+    function getVestedAmount(uint256 vestingId_) public view virtual returns (uint256) {
+        VestingData storage _vesting = _vestings[vestingId_];
+        Schedule storage _schedule = _schedules[_vesting.scheduleId];
+
+        return _getVestedAmount(_vesting, _schedule, block.timestamp);
+    }
+
+    function _getWithdrawableAmount(
+        VestingData storage vesting,
+        Schedule storage schedule,
+        uint256 timestamp_
+    ) internal view virtual returns (uint256) {
+        return
+            _vestingCalculation(
+                schedule,
+                vesting.vestingAmount,
+                vesting.vestingStartTime,
+                timestamp_
+            ) - vesting.paidAmount;
+    }
+
+    function _getVestedAmount(
+        VestingData storage vesting,
+        Schedule storage schedule,
+        uint256 timestamp_
+    ) internal view virtual returns (uint256) {
+        return
+            _vestingCalculation(
+                schedule,
+                vesting.vestingAmount,
+                vesting.vestingStartTime,
+                timestamp_
+            );
+    }
+
+    function _createVesting(VestingData memory vesting_) internal virtual returns (uint256) {
+        require(
+            vesting_.vestingStartTime > 0,
+            "VestingWallet: cannot create vesting for zero time"
+        );
+        require(
+            vesting_.vestingAmount > 0,
+            "VestingWallet: cannot create vesting for zero amount"
+        );
+        require(
+            vesting_.beneficiary != address(0),
+            "VestingWallet: cannot create vesting for zero address"
+        );
+
+        Schedule storage _schedule = _schedules[vesting_.scheduleId];
+
+        require(
+            vesting_.vestingStartTime +
+                _schedule.scheduleData.durationInPeriods *
+                _schedule.scheduleData.secondsInPeriod >
+                block.timestamp,
+            "VestingWallet: cannot create vesting for past date"
+        );
+
+        _vestingId++;
+
+        _beneficiaryIds[vesting_.beneficiary].add(_vestingId);
+
+        _vestings[_vestingId] = vesting_;
+
+        emit VestingCreated(_vestingId, vesting_.beneficiary, vesting_.vestingToken);
+
+        return _vestingId;
+    }
+
+    function _withdrawFromVesting(
```

Comment:
We need to transfer the token don't we?

Reply:
Didn't we agree not to use token transfers in this contract? I implemented transfer in mock contract

Reply:
But what's the matter with withdrawals? We did agree with vesting creation as it is a bit ambiguous (creators may want to airdrop the contract).

Reply:
And I am not sure if this function has to be internal at all

Reply:
Also it would be wise to check if vesting actually exists